test_runner/batch_others/test_proxy.py

```diff
@@ -22,6 +26,99 @@ def test_password_hack(static_proxy):
+    # Expected output:
+    #   A notice that contains a welcome to neon message that contains a http url in the form of
+    #   "http://dummy-uri/<psql_session_id>"
+    #   (here dummy-uri was passed from the neon_fixture that created the link_proxy
+    #   The url has to be on its own line.
+    #   Example:
+    #   "
+    #   NOTICE:  Welcome to Neon!
+    #   Authenticate by visiting:
+    #      http://dummy-uri/<psql_session_id>
+    #   "
+
+    max_num_lines_of_welcome_message = 15
+    attempt = 0
+    psql_session_id = None
+
+    log.info(f"Neon Welcome Message:")
+    for attempt in range(max_num_lines_of_welcome_message):
+        raw_line = await proc.stderr.readline()
+        if not raw_line:
+            # output ended
+            break
+        log.info(f"{raw_line}")
+        line = raw_line.decode("utf-8").strip()
+        if line.startswith("http"):
+            line_parts = line.split("/")
+            psql_session_id = line_parts[-1]
+            link_auth_uri = '/'.join(line_parts[:-1])
+            assert link_auth_uri == link_proxy.link_auth_uri, \
+                f"Line='{line}' should contain a http auth link of form '{link_proxy.link_auth_uri}/<psql_session_id>'."
+            break
+        log.debug("line %d does not contain expected result: %s", attempt, line)
+
+    assert attempt <= max_num_lines_of_welcome_message, "exhausted given attempts, did not get the result"
+    assert psql_session_id is not None, "psql_session_id not found from output of proc.stderr.readline()"
+    log.debug(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
+    log.info(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
+
+    # make a db and send to proxy.
+    # greate a postgres instance and give the proxy the info to send queries to it.
+
+    pg_user = "proxy"
+    pg_password = "password"
+
+    vanilla_pg.start()
+    vanilla_pg.safe_psql("create user "+pg_user+" with login superuser password '"+pg_password+"'")
+
+    port = vanilla_pg.default_options['port']
+    host = vanilla_pg.default_options['host']
+    dbname = vanilla_pg.default_options['dbname']
+
+    cmd_line_args__to__mgmt = [
+        "psql",
+        "-h",
+        "127.0.0.1",
+        "-p",
+        "7000",
+        '-c',
+        '{"session_id": "' + str(psql_session_id) + '"'+
+        ',"result":{"Success":{"host":"'+host+'","port":'+str(port)+
+        ',"dbname":"'+dbname+'"'+
+        ',"user":"'+pg_user+'"'+
+        ',"password":"'+pg_password+'"}}}'
+    ]
+
+    for arg_id, arg in enumerate(cmd_line_args__to__mgmt):
+        log.info(f"arg_id={arg_id}: {arg}")
+
+    log.info(f"running cmd line: {cmd_line_args__to__mgmt}")
+    p = subprocess.Popen(cmd_line_args__to__mgmt, stdout=subprocess.PIPE)
+    out, err = p.communicate()
+    log.info(f"output of running cmd line: out={out}; err={err}")
+
+    link_proxy.safe_psql('select 1', options='project=generic-project-name')
```

Comment:
Gives error: FAILED test_runner/batch_others/test_proxy.py::test_psql_session_id - psycopg2.OperationalError: connection to server on socket "/tmp/.s.PGSQL.15001" failed: No such file or directory

Reply:
```
full output: 
======================================================================================================= FAILURES =======================================================================================================
_________________________________________________________________________________________________ test_psql_session_id _________________________________________________________________________________________________

vanilla_pg = <fixtures.neon_fixtures.VanillaPostgres object at 0x10fc730d0>, link_proxy = <fixtures.neon_fixtures.NeonProxy object at 0x10fc73370>

    @pytest.mark.asyncio
    async def test_psql_session_id(vanilla_pg, link_proxy):
        """
        Test copied and modified from: test_project_psql_link_auth test from cloud/tests_e2e/tests/test_project.py
        This is only one half of the test: the half that establishes connection to the proxy and retrieves the session_id
        The second half of the test required more heavy machinery to be ported from the cloud codebase
            (such as: project_state(..), and psql_project_connect)
        to connect to postgres and make sure the end-to-end behavior is satisfied.
        """
        psql = PSQL(
            host=link_proxy.host,
            port=link_proxy.proxy_port,
        )
        proc = await psql.run()
    
        # Read line by line output of psql.
        # We expect to see result in first 3 lines. [this is flexible, change if need be]
        # Expected output:
        #   A notice that contains a welcome to neon message that contains a http url in the form of
        #   "http://dummy-uri/<psql_session_id>"
        #   (here dummy-uri was passed from the neon_fixture that created the link_proxy
        #   The url has to be on its own line.
        #   Example:
        #   "
        #   NOTICE:  Welcome to Neon!
        #   Authenticate by visiting:
        #      http://dummy-uri/<psql_session_id>
        #   "
    
        max_num_lines_of_welcome_message = 15
        attempt = 0
        psql_session_id = None
    
        log.info(f"Neon Welcome Message:")
        for attempt in range(max_num_lines_of_welcome_message):
            raw_line = await proc.stderr.readline()
            if not raw_line:
                # output ended
                break
            log.info(f"{raw_line}")
            line = raw_line.decode("utf-8").strip()
            if line.startswith("http"):
                line_parts = line.split("/")
                psql_session_id = line_parts[-1]
                link_auth_uri = '/'.join(line_parts[:-1])
                assert link_auth_uri == link_proxy.link_auth_uri, \
                    f"Line='{line}' should contain a http auth link of form '{link_proxy.link_auth_uri}/<psql_session_id>'."
                break
            log.debug("line %d does not contain expected result: %s", attempt, line)
    
        assert attempt <= max_num_lines_of_welcome_message, "exhausted given attempts, did not get the result"
        assert psql_session_id is not None, "psql_session_id not found from output of proc.stderr.readline()"
        log.debug(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
        log.info(f"proc.stderr.readline() #{attempt} has the result: {psql_session_id=}")
    
        # make a db and send to proxy.
        # greate a postgres instance and give the proxy the info to send queries to it.
    
        pg_user = "proxy"
        pg_password = "password"
    
        vanilla_pg.start()
        vanilla_pg.safe_psql("create user "+pg_user+" with login superuser password '"+pg_password+"'")
    
        port = vanilla_pg.default_options['port']
        host = vanilla_pg.default_options['host']
        dbname = vanilla_pg.default_options['dbname']
    
        cmd_line_args__to__mgmt = [
            "psql",
            "-h",
            "127.0.0.1",
            "-p",
            "7000",
            '-c',
            '{"session_id": "' + str(psql_session_id) + '"'+
            ',"result":{"Success":{"host":"'+host+'","port":'+str(port)+
            ',"dbname":"'+dbname+'"'+
            ',"user":"'+pg_user+'"'+
            ',"password":"'+pg_password+'"}}}'
        ]
    
        for arg_id, arg in enumerate(cmd_line_args__to__mgmt):
            log.info(f"arg_id={arg_id}: {arg}")
    
        log.info(f"running cmd line: {cmd_line_args__to__mgmt}")
        p = subprocess.Popen(cmd_line_args__to__mgmt, stdout=subprocess.PIPE)
        out, err = p.communicate()
        log.info(f"output of running cmd line: out={out}; err={err}")
    
>       link_proxy.safe_psql('select 1', options='project=generic-project-name')

test_runner/batch_others/test_proxy.py:119: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_runner/fixtures/neon_fixtures.py:397: in safe_psql
    return self.safe_psql_many([query], **kwargs)[0]
test_runner/fixtures/neon_fixtures.py:405: in safe_psql_many
    with closing(self.connect(**kwargs)) as conn:
test_runner/fixtures/neon_fixtures.py:316: in connect
    conn = psycopg2.connect(**self.conn_options(**kwargs))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

dsn = "port=15001 options='-cstatement_timeout=120s project=generic-project-name'", connection_factory = None, cursor_factory = None
kwargs = {'options': '-cstatement_timeout=120s project=generic-project-name', 'port': 15001}, kwasync = {}

    def connect(dsn=None, connection_factory=None, cursor_factory=None, **kwargs):
        """
        Create a new database connection.
    
        The connection parameters can be specified as a string:
    
            conn = psycopg2.connect("dbname=test user=postgres password=secret")
    
        or using a set of keyword arguments:
    
            conn = psycopg2.connect(database="test", user="postgres", password="secret")
    
        Or as a mix of both. The basic connection parameters are:
    
        - *dbname*: the database name
        - *database*: the database name (only as keyword argument)
        - *user*: user name used to authenticate
        - *password*: password used to authenticate
        - *host*: database host address (defaults to UNIX socket if not provided)
        - *port*: connection port number (defaults to 5432 if not provided)
    
        Using the *connection_factory* parameter a different class or connections
        factory can be specified. It should be a callable object taking a dsn
        argument.
    
        Using the *cursor_factory* parameter, a new default cursor factory will be
        used by cursor().
    
        Using *async*=True an asynchronous connection will be created. *async_* is
        a valid alias (for Python versions where ``async`` is a keyword).
    
        Any other keyword parameter will be passed to the underlying client
        library: the list of supported parameters depends on the library version.
    
        """
        kwasync = {}
        if 'async' in kwargs:
            kwasync['async'] = kwargs.pop('async')
        if 'async_' in kwargs:
            kwasync['async_'] = kwargs.pop('async_')
    
        dsn = _ext.make_dsn(dsn, **kwargs)
>       conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
E       psycopg2.OperationalError: connection to server on socket "/tmp/.s.PGSQL.15001" failed: No such file or directory
E               Is the server running locally and accepting connections on that socket?

../../Library/Caches/pypoetry/virtualenvs/zenith-gI5I-K-t-py3.9/lib/python3.9/site-packages/psycopg2/__init__.py:122: OperationalError
------------------------------------------------------------------------------------------------ Captured stdout setup -------------------------------------------------------------------------------------------------
Version: git:0e583fde4baffbd989e7389a6370b238b5d71d5a
Authentication backend: Link
Starting http on 127.0.0.1:15000
Starting mgmt on 127.0.0.1:7000
Starting proxy on 127.0.0.1:15001
-------------------------------------------------------------------------------------------------- Captured log setup --------------------------------------------------------------------------------------------------
2022-08-19 02:12:49.765 INFO [neon_fixtures.py:2338] get_test_output_dir is /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id
2022-08-19 02:12:49.766 INFO [neon_fixtures.py:2359] test_output_dir is /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id
2022-08-19 02:12:49.847 INFO [neon_fixtures.py:1627] Running command "/Users/klimentserafimov/CLionProjects/neon/tmp_install/bin/initdb -D /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id/pgdata-vanilla"
2022-08-19 02:12:49.848 INFO [utils.py:38] Capturing stdout to "initdb_1.stdout" and stderr to "initdb_1.stderr"
2022-08-19 02:12:51.9   INFO [neon_fixtures.py:1811] starting proxy with command line ::: /Users/klimentserafimov/CLionProjects/neon/target/debug/proxy --http 127.0.0.1:15000 --proxy 127.0.0.1:15001 --auth-backend link --uri http://dummy-uri
------------------------------------------------------------------------------------------------- Captured stdout call -------------------------------------------------------------------------------------------------
accepted connection from 127.0.0.1:50502
got message: SslRequest
got message: StartupMessage { major_version: 3, minor_version: 0, params: {"": "", "user": "klimentserafimov", "project": "generic-project-name", "application_name": "psql", "database": "main"} }
accepted connection from 127.0.0.1:50504
Got mgmt query [redacted]
-------------------------------------------------------------------------------------------------- Captured log call ---------------------------------------------------------------------------------------------------
2022-08-19 02:12:51.23  INFO [neon_fixtures.py:1761] running psql with command line ::: psql postgres://127.0.0.1:15001/main?options=project%3Dgeneric-project-name
2022-08-19 02:12:51.27  INFO [test_proxy.py:62] Neon Welcome Message:
2022-08-19 02:12:51.80  INFO [test_proxy.py:68] b'NOTICE:  Welcome to Neon!\n'
2022-08-19 02:12:51.81  INFO [test_proxy.py:68] b'Authenticate by visiting:\n'
2022-08-19 02:12:51.81  INFO [test_proxy.py:68] b'    http://dummy-uri/b101f655b5af5e34\n'
2022-08-19 02:12:51.81  INFO [test_proxy.py:82] proc.stderr.readline() #2 has the result: psql_session_id='b101f655b5af5e34'
2022-08-19 02:12:51.81  INFO [neon_fixtures.py:1627] Running command "/Users/klimentserafimov/CLionProjects/neon/tmp_install/bin/pg_ctl -w -D /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id/pgdata-vanilla -l /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id/pgdata-vanilla/pg.log start"
2022-08-19 02:12:51.82  INFO [utils.py:38] Capturing stdout to "pg_ctl_2.stdout" and stderr to "pg_ctl_2.stderr"
2022-08-19 02:12:51.215 INFO [neon_fixtures.py:408] Executing query: create user proxy with login superuser password 'password'
2022-08-19 02:12:51.231 INFO [test_proxy.py:112] arg_id=0: psql
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=1: -h
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=2: 127.0.0.1
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=3: -p
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=4: 7000
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=5: -c
2022-08-19 02:12:51.232 INFO [test_proxy.py:112] arg_id=6: {"session_id": "b101f655b5af5e34","result":{"Success":{"host":"localhost","port":15000,"dbname":"postgres","user":"proxy","password":"password"}}}
2022-08-19 02:12:51.232 INFO [test_proxy.py:114] running cmd line: ['psql', '-h', '127.0.0.1', '-p', '7000', '-c', '{"session_id": "b101f655b5af5e34","result":{"Success":{"host":"localhost","port":15000,"dbname":"postgres","user":"proxy","password":"password"}}}']
2022-08-19 02:12:51.242 INFO [test_proxy.py:117] output of running cmd line: out=b' data \n------\n ok\n(1 row)\n\n'; err=None
------------------------------------------------------------------------------------------------ Captured log teardown -------------------------------------------------------------------------------------------------
2022-08-19 02:12:51.385 INFO [neon_fixtures.py:1627] Running command "/Users/klimentserafimov/CLionProjects/neon/tmp_install/bin/pg_ctl -w -D /Users/klimentserafimov/CLionProjects/neon/test_output/test_psql_session_id/pgdata-vanilla stop"
2022-08-19 02:12:51.385 INFO [utils.py:38] Capturing stdout to "pg_ctl_3.stdout" and stderr to "pg_ctl_3.stderr"
-------------------------------------------------------------------------------------------------- Benchmark results ---------------------------------------------------------------------------------------------------
/Users/klimentserafimov/CLionProjects/neon/test_runner/fixtures/benchmark_fixture.py:437: UserWarning: no out dir provided to store performance test results
  warnings.warn("no out dir provided to store performance test results")
=============================================================================================== short test summary info ================================================================================================
FAILED test_runner/batch_others/test_proxy.py::test_psql_session_id - psycopg2.OperationalError: connection to server on socket "/tmp/.s.PGSQL.15001" failed: No such file or directory
================================================================================================== 1 failed in 1.76s ===================================================================================================
Exception ignored in: <function BaseSubprocessTransport.__del__ at 0x10e8f8820>
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_subprocess.py", line 126, in __del__
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_subprocess.py", line 104, in close
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/unix_events.py", line 536, in close
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/unix_events.py", line 560, in _close
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 751, in call_soon
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 515, in _check_closed
RuntimeError: Event loop is closed

```